upgrade fail under windows

##### **Overview of the issue**

I tried to upgrade my code to 4.6.2 but encountered error:

$ jhipster upgrade
Executing jhipster:upgrade
Options:
Welcome to the JHipster Upgrade Sub-Generator
This will upgrade your current application codebase to the latest JHipster version
Looking for latest generator-jhipster version...
New generator-jhipster version found: 4.6.2
Git repository detected
Checked out branch "jhipster_upgrade"
Updating generator-jhipster to 4.6.2 . This might take some time...
Updated generator-jhipster to version 4.6.2
Cleaned up directory
Regenerating application with JHipster 4.6.2...
events.js:160
      throw er; // Unhandled 'error' event
      ^

Error: ERROR! Something went wrong while generating project! '$' is not recognized as an internal or external command,
operable program or batch file.

    at Environment.error (C:\Users\Powerful User\AppData\Local\Yarn\config\global\node_modules\yeoman-environment\lib\environment.js:140:40)
    at constructor.error (C:\Users\Powerful User\AppData\Local\Yarn\config\global\node_modules\generator-jhipster\generators\generator-base.js:1610:18)
    at shelljs.exec (C:\Users\Powerful User\AppData\Local\Yarn\config\global\node_modules\generator-jhipster\generators\upgrade\index.js:86:23)
    at C:\Users\Powerful User\AppData\Local\Yarn\config\global\node_modules\shelljs\src\exec.js:204:9
    at ChildProcess.exithandler (child_process.js:211:5)
    at emitTwo (events.js:106:13)
    at ChildProcess.emit (events.js:191:7)
    at maybeClose (internal/child_process.js:886:16)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:226:5)


##### **Motivation for or Use Case**

To help who encounter same issue under windows when upgrading a jhipster generated project

##### **Reproduce the error**

Step 1: 
yarn global upgrade generator-jhipster

Step 2:
jhipster upgrade

This error occurs in power shell AND git bash.

##### **Related issues**

None

##### **Suggest a Fix**

<!-- For bug reports, if you can't fix the bug yourself, perhaps you can point to what might be
  causing the problem (line of code or commit) -->

##### **JHipster Version(s)**

Upgraded jhipster generator to 4.6.2 then try to upgrade my code base from 4.5.6 to 4.6.2

##### **JHipster configuration**

$ jhipster info
Executing jhipster:info
Options:
Welcome to the JHipster Information Sub-Generator

##### **JHipster Version(s)**

```
C:\code\sibodell-hse\hse-backend
`-- generator-jhipster@4.6.2

```


##### **JHipster configuration, a `.yo-rc.json` file generated in the root folder**


<details>
<summary>.yo-rc.json file</summary>
<pre>
{
  "generator-jhipster": {
    "promptValues": {
      "packageName": "com.bigcompany.hse",
      "nativeLanguage": "zh-cn"
    },
    "jhipsterVersion": "4.5.6",
    "baseName": "hse",
    "packageName": "com.bigcompany.hse",
    "packageFolder": "com/bigcompany/hse",
    "serverPort": "8080",
    "authenticationType": "jwt",
    "hibernateCache": "ehcache",
    "clusteredHttpSession": false,
    "websocket": false,
    "databaseType": "sql",
    "devDatabaseType": "mysql",
    "prodDatabaseType": "mysql",
    "searchEngine": false,
    "messageBroker": false,
    "serviceDiscoveryType": false,
    "buildTool": "maven",
    "enableSocialSignIn": false,
    "jwtSecretKey": "replaced-by-jhipster-info",
    "clientFramework": "angular1",
    "useSass": false,
    "clientPackageManager": "yarn",
    "applicationType": "monolith",
    "testFrameworks": [],
    "jhiPrefix": "jhi",
    "enableTranslation": true,
    "nativeLanguage": "zh-cn",
    "languages": [
      "zh-cn",
      "en"
    ]
  }
}
</pre>
</details>


##### **JDL for the Entity configuration(s) `entityName.json` files generated in the `.jhipster` directory**

ls: no such file or directory: .jhipster/*.json
<details>
<summary>JDL entity definitions</summary>

<pre>





</pre>
</details>


##### **Environment and Tools**

java version "1.8.0_141"
Java(TM) SE Runtime Environment (build 1.8.0_141-b15)
Java HotSpot(TM) 64-Bit Server VM (build 25.141-b15, mixed mode)

git version 2.12.2.windows.2

node: v6.10.2

npm: 5.0.4

bower: 1.8.0

gulp:
[12:11:04] CLI version 1.3.0
[12:11:04] Local version 3.9.1

yeoman: 2.0.0

yarn: 0.27.5

Docker version 1.12.5, build 7392c3b

docker-compose version 1.9.0, build 2585387

Execution complete

##### **Entity configuration(s) `entityName.json` files generated in the `.jhipster` directory**

NONE (We generated some entities then moved the .json files to other directory, this bug should have nothing to do with entities)

##### **Browsers and Operating System**

Windows 10 Pro

- [* ] Checking this box is mandatory (yes, I've read everything)

@iconben could you try to reproduce with shelljs in verbose mode? Just edit `generator-jhipster\generators\upgrade\index.js` and set verbose mode:

~~~js
const shelljs = require('shelljs');
shelljs.config.verbose = true;
~~~

I tried to reproduce but I have a more basic issue first about tabtab

I think we should always set shelljs in verbose mode, just compare the 2 outputs below:

<details>
<summary>shelljs in normal mode</summary>
<pre>
jhipster upgrade
Executing jhipster:upgrade
Welcome to the JHipster Upgrade Sub-Generator
This will upgrade your current application codebase to the latest JHipster version
Looking for latest generator-jhipster version...
New generator-jhipster version found: 4.6.2
Git repository detected
WARNING!  local changes found.
        Please commit/stash them before upgrading
events.js:160
      throw er; // Unhandled 'error' event
      ^

Error: [31m[1mERROR![22m[39m Exiting process
    at Environment.error (C:\Users\Gaël\AppData\Local\Yarn\config\global\node_modules\generator-jhipster\node_modules\yeoman-environment\lib\environment.js:140:40)
    at constructor.error (C:\Users\Gaël\AppData\Local\Yarn\config\global\node_modules\generator-jhipster\generators\generator-base.js:1645:18)
    at gitExec (C:\Users\Gaël\AppData\Local\Yarn\config\global\node_modules\generator-jhipster\generators\upgrade\index.js:179:26)
    at C:\Users\Gaël\AppData\Local\Yarn\config\global\node_modules\generator-jhipster\node_modules\shelljs\src\exec.js:199:9
    at ChildProcess.exithandler (child_process.js:195:7)
    at emitTwo (events.js:106:13)
    at ChildProcess.emit (events.js:191:7)
    at maybeClose (internal/child_process.js:886:16)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:226:5)
</pre>
</details>

<details>
<summary>shelljs in verbose mode</summary>
<pre>
jhipster upgrade
Executing jhipster:upgrade
Welcome to the JHipster Upgrade Sub-Generator
This will upgrade your current application codebase to the latest JHipster version
Looking for latest generator-jhipster version...
New generator-jhipster version found: 4.6.2
Git repository detected
Created branch jhipster_upgrade
Cleaned up directory
Installing JHipster 4.6.1 locally
events.js:160
      throw er; // Unhandled 'error' event
      ^

Error: [31m[1mERROR![22m[39m Something went wrong while installing the JHipster generator! yarn add v0.27.5
info No lockfile found.
[1/4] Resolving packages...
[2/4] Fetching packages...
[3/4] Linking dependencies...
[4/4] Building fresh packages...
info Visit https://yarnpkg.com/en/docs/cli/add for documentation about this command.
 error E:\projets\issues\upgrade\node_modules\generator-jhipster: Command failed.
Exit code: 1
Command: C:\WINDOWS\system32\cmd.exe
Arguments: /d /s /c tabtab install --name jhipster --auto
Directory: E:\projets\issues\upgrade\node_modules\generator-jhipster
Output:
module.js:471
    throw err;
    ^

Error: Cannot find module 'E:\projets\issues\upgrade\node_modules\issues\upgrade\node_modules\tabtab\bin\tabtab'
    at Function.Module._resolveFilename (module.js:469:15)
    at Function.Module._load (module.js:417:25)
    at Module.runMain (module.js:604:10)
    at run (bootstrap_node.js:393:7)
    at startup (bootstrap_node.js:150:9)
    at bootstrap_node.js:508:3

    at Environment.error (C:\Users\Gaël\AppData\Local\Yarn\config\global\node_modules\generator-jhipster\node_modules\yeoman-environment\lib\environment.js:140:40)
    at constructor.error (C:\Users\Gaël\AppData\Local\Yarn\config\global\node_modules\generator-jhipster\generators\generator-base.js:1645:18)
    at shelljs.exec (C:\Users\Gaël\AppData\Local\Yarn\config\global\node_modules\generator-jhipster\generators\upgrade\index.js:226:31)
    at C:\Users\Gaël\AppData\Local\Yarn\config\global\node_modules\generator-jhipster\node_modules\shelljs\src\exec.js:204:9
    at ChildProcess.exithandler (child_process.js:211:5)
    at emitTwo (events.js:106:13)
    at ChildProcess.emit (events.js:191:7)
    at maybeClose (internal/child_process.js:886:16)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:226:5)
</pre>
</details>

--- a/generators/upgrade/index.js
+++ b/generators/upgrade/index.js
@@ -79,7 +79,8 @@ module.exports = UpgradeGenerator.extend({
         this.log(`Regenerating application with JHipster ${version}...`);
         let generatorCommand = 'yo jhipster';
         if (semver.gte(version, FIRST_CLI_SUPPORTED_VERSION)) {
-            generatorCommand = this.clientPackageManager === 'yarn' ? '$(yarn bin)/jhipster' : '$(npm bin)/jhipster';
+            const generatorDir = this.clientPackageManager === 'yarn' ? shelljs.exec('yarn bin') : shelljs.exec('npm bin');
+            generatorCommand = `${generatorDir.replace('\n', '')}/jhipster`;
         }
         shelljs.exec(`${generatorCommand} --with-entities --force --skip-install`, { silent: this.silent }, (code, msg, err) => {
             if (code === 0) this.log(chalk.green(`Successfully regenerated application with JHipster ${version}`));

--- a/generators/upgrade/index.js
+++ b/generators/upgrade/index.js
@@ -79,7 +79,7 @@ module.exports = UpgradeGenerator.extend({
         this.log(`Regenerating application with JHipster ${version}...`);
         let generatorCommand = 'yo jhipster';
         if (semver.gte(version, FIRST_CLI_SUPPORTED_VERSION)) {
-            const generatorDir = this.clientPackageManager === 'yarn' ? shelljs.exec('yarn bin', { silent: true }).stdout : shelljs.exec('npm bin', { silent: true }).stdout;
+            const generatorDir = this.clientPackageManager === 'yarn' ? shelljs.exec('yarn bin', { silent: this.silent }).stdout : shelljs.exec('npm bin', { silent: this.silent }).stdout;
             generatorCommand = `${generatorDir.replace('\n', '')}/jhipster`;
         }
         shelljs.exec(`${generatorCommand} --with-entities --force --skip-install`, { silent: this.silent }, (code, msg, err) => {

--- a/generators/upgrade/index.js
+++ b/generators/upgrade/index.js
@@ -79,7 +79,7 @@ module.exports = UpgradeGenerator.extend({
         this.log(`Regenerating application with JHipster ${version}...`);
         let generatorCommand = 'yo jhipster';
         if (semver.gte(version, FIRST_CLI_SUPPORTED_VERSION)) {
-            const generatorDir = this.clientPackageManager === 'yarn' ? shelljs.exec('yarn bin') : shelljs.exec('npm bin');
+            const generatorDir = this.clientPackageManager === 'yarn' ? shelljs.exec('yarn bin', { silent: true }).stdout : shelljs.exec('npm bin', { silent: true }).stdout;
             generatorCommand = `${generatorDir.replace('\n', '')}/jhipster`;
         }
         shelljs.exec(`${generatorCommand} --with-entities --force --skip-install`, { silent: this.silent }, (code, msg, err) => {

